File: backend/api/routes/providers.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request, status

from backend.api.dependencies import get_provider_service
from backend.api.schemas.providers import (
    ProviderDraft,
    ProviderModelSelectionRequest,
    ProviderModelSelectionResult,
    ProviderModelVerifyRequest,
    ProviderModelVerifyResult,
    ProviderPublic,
    ProviderSlug,
    ProviderTestResult,
    ProviderUpdate,
)
from backend.application.provider_clients import ProviderConnectionError
from backend.application.provider_service import (
    ProviderConfigurationError,
    ProviderService,
)
from backend.security.auth import require_admin


def require_loopback(request: Request) -> None:
    host = request.client.host if request.client else ""
    if host not in {"127.0.0.1", "::1", "test"}:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={
                "code": "SETTINGS_LOCAL_ONLY",
                "message": "API settings are available only on this computer",
                "retryable": False,
            },
        )


router = APIRouter(
    prefix="/settings/providers",
    tags=["provider-settings"],
    dependencies=[Depends(require_loopback), Depends(require_admin)],
)


def _raise_provider_error(error: Exception) -> None:
    if isinstance(error, (ProviderConfigurationError, ProviderConnectionError)):
        raise HTTPException(
            status_code=(
                status.HTTP_503_SERVICE_UNAVAILABLE
                if error.retryable
                else status.HTTP_422_UNPROCESSABLE_CONTENT
            ),
            detail={
                "code": error.code,
                "message": error.message,
                "retryable": error.retryable,
            },
        ) from error
    raise error


@router.get("", response_model=list[ProviderPublic])
async def list_providers(
    service: ProviderService = Depends(get_provider_service),
) -> list[ProviderPublic]:
    return await service.list_public()


@router.post("/{slug}/test", response_model=ProviderTestResult)
async def test_provider(
    slug: ProviderSlug,
    draft: ProviderDraft,
    service: ProviderService = Depends(get_provider_service),
) -> ProviderTestResult:
    try:
        result = await service.test_draft(slug, draft)
    except (ProviderConfigurationError, ProviderConnectionError) as error:
        _raise_provider_error(error)
    return ProviderTestResult(message=result.message, models=list(result.models))


@router.put("/{slug}", response_model=ProviderPublic)
async def update_provider(
    slug: ProviderSlug,
    update: ProviderUpdate,
    service: ProviderService = Depends(get_provider_service),
) -> ProviderPublic:
    try:
        return await service.save(slug, update)
    except (ProviderConfigurationError, ProviderConnectionError) as error:
        _raise_provider_error(error)
        raise AssertionError("unreachable")


@router.post("/{slug}/models/verify", response_model=ProviderModelVerifyResult)
async def verify_provider_model(
    slug: ProviderSlug,
    request: ProviderModelVerifyRequest,
    service: ProviderService = Depends(get_provider_service),
) -> ProviderModelVerifyResult:
    try:
        return await service.verify_model(
            slug,
            request.model,
            set_default=request.set_default,
            is_automatic=request.is_automatic,
        )
    except (ProviderConfigurationError, ProviderConnectionError) as error:
        _raise_provider_error(error)
        raise AssertionError("unreachable")


@router.patch(
    "/{slug}/models/{model}/selection",
    response_model=ProviderModelSelectionResult,
)
async def select_provider_model(
    slug: ProviderSlug,
    model: str,
    request: ProviderModelSelectionRequest,
    service: ProviderService = Depends(get_provider_service),
) -> ProviderModelSelectionResult:
    try:
        return await service.set_model_selected(slug, model, request.is_selected)
    except (ProviderConfigurationError, ProviderConnectionError) as error:
        _raise_provider_error(error)
        raise AssertionError("unreachable")


__all__ = ["require_loopback", "router"]

File: frontend/lib/api/providers.ts
import { apiFetch } from "./client";
import type {
  ProviderConfiguration,
  ProviderDraftPayload,
  ProviderModelVerifyResult,
  ProviderModelSelectionResult,
  ProviderSlug,
  ProviderTestResult,
  ProviderUpdatePayload,
} from "./types";

const RETIRED_PROVIDER_SLUGS = new Set(["cattoken", "cattoken_claude"]);

/**
 * Provider CRUD is scoped: global `/settings/providers` (admin's default set)
 * when no groupId is given, else the ApiGroup's own provider set under
 * `/admin/api-groups/{groupId}/providers` (admin only).
 */
function providerBase(groupId?: string): string {
  return groupId
    ? `/admin/api-groups/${groupId}/providers`
    : "/settings/providers";
}

export async function listProviders(
  groupId?: string
): Promise<ProviderConfiguration[]> {
  const providers = await apiFetch<Array<ProviderConfiguration & { slug: string }>>(
    providerBase(groupId)
  );
  return providers.filter(
    (provider): provider is ProviderConfiguration =>
      !RETIRED_PROVIDER_SLUGS.has(provider.slug)
  );
}

export function verifyProviderModel(
  slug: ProviderSlug,
  model: string,
  setDefault = false,
  isAutomatic = false,
  groupId?: string
): Promise<ProviderModelVerifyResult> {
  return apiFetch<ProviderModelVerifyResult>(
    `${providerBase(groupId)}/${slug}/models/verify`,
    {
      method: "POST",
      body: {
        model,
        set_default: setDefault,
        ...(isAutomatic ? { is_automatic: true } : {}),
      },
    }
  );
}

export function selectProviderModel(
  slug: ProviderSlug,
  model: string,
  isSelected: boolean,
  groupId?: string
): Promise<ProviderModelSelectionResult> {
  return apiFetch<ProviderModelSelectionResult>(
    `${providerBase(groupId)}/${slug}/models/${encodeURIComponent(model)}/selection`,
    { method: "PATCH", body: { is_selected: isSelected } }
  );
}

export function testProvider(
  slug: ProviderSlug,
  payload: ProviderDraftPayload,
  groupId?: string
): Promise<ProviderTestResult> {
  return apiFetch<ProviderTestResult>(`${providerBase(groupId)}/${slug}/test`, {
    method: "POST",
    body: payload,
  });
}

export function updateProvider(
  slug: ProviderSlug,
  payload: ProviderUpdatePayload,
  groupId?: string
): Promise<ProviderConfiguration> {
  return apiFetch<ProviderConfiguration>(`${providerBase(groupId)}/${slug}`, {
    method: "PUT",
    body: payload,
  });
}
